### ISSUE TITLE
Supplementary Report - Create a Review Changes page [New Feature]

### ISSUE THREAD

**Issue by zoeyli-46** [bcgov/cas-reporting#479]
#### Description:
There is a new Review Changes page in supplementary report mode, which comes before the Final Review page
- [ ] [Review Changes wireframe](https://www.figma.com/design/LsDmLDCdnJqI8UahyqvOD8/BCIERS-Designs?node-id=27418-5212&t=XJfevJ7Ha4Eekqpn-4)

See also #480 

#### Acceptance Criteria:

Given I am on the last page of additional report data section,
When I click continue
Then I am brought to the Review Changes page 
(The review changes page is now the first page in the sign off and submit section)

<img width="162" height="110" alt="Image" src="https://github.com/user-attachments/assets/a95bd9f0-d681-444b-88a6-ee72eebafc5a" />

Given I am on the Review Changes page, I see the following:
- [ ] the page section and fields that were changed, in view-only mode
- [ ] The dif shows the old value with a strike through, and the new value in bold 
- [ ] Changes to the report information fields will show all of the headers and subheaders that precede the field that was changed. i.e., if an emissions field was changed, then the facility name, the activity name, fuel, and emission title is displayed. Screenshot below as example.

<img width="385" alt="Image" src="https://github.com/user-attachments/assets/66dd7611-ddb3-44ce-b1e2-a7afa67609c5" />

- [x] 1 editable text box "reason for edits" (mandatory field)

NOTE FOR DEVS: 

- This page will need to be renamed "Review Changes" page
- The "reason for error" editable text box should have been completed in ticket #729 
-  When users click 'save & continue' they are taken to the Final Review page. We have this set up to show the "Reason for error" text box at the top of the final review page. We do not intend to show the "difs" on this page, so it will just display the rest of the report as is. 
- the diff should not be shown if the registration purpose has changed between report versions

#### Development Checklist:

- [ ] Define route: GET /reports/:reportId/review-changes
- [ ] Fetch the [final review data](https://github.com/bcgov/cas-reporting/issues/535) of the last two report versions and return only the diff in the format:
tableName: { fieldName: { old, new } }
- [ ] Create the Review Changes page which takes the diff data and displays the field names and values changed between the last two versions
- [ ] Also add the "reason for error" field, as implemented in [#729](https://github.com/bcgov/cas-reporting/issues/729)
- [ ] Add required tests
- [ ] Meets the DOD

Example Output:
`{
  "report_operation": {
    "operator_legal_name": {
      "old": "abc",
      "new": "def"
    }
  },
  "facility_report": {
    "facility_name": {
      "old": "Old Facility",
      "new": "New Facility"
    }
  }
}
`

#### Definition of Ready (Note: If any of these points are not applicable, mark N/A)

- [ ] User story is included
- [ ] User role and type are identified
- [ ] Acceptance criteria are included
- [ ] Wireframes are included (if required)
- [ ] Design / Solution is accepted by Product Owner
- [ ] Dependencies are identified (technical, business, regulatory/policy)
- [ ] Story has been estimated (under 13 pts)

#### Definition of Done (Note: If any of these points are not applicable, mark N/A)

- [ ] Acceptance criteria are tested by the CI pipeline
- [ ] UI meets accessibility requirements
- [ ] Configuration changes are documented, documentation and designs are updated
- [ ] Passes code peer-review
- [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
- [ ] Ticket is ready to be merged to main branch
- [ ] Can be demoed in Sprint Review
- [ ] Bugs or future work cards are identified and created
- [ ] Reviewed and approved by Product Owner

#### Notes:

- Review code in CIF that was used for similar work?

**Comment by zoeyli-46** [comment]
@patriciarussellCAS new supplementary report ticket for the Review Changes page

**Comment by patriciarussellCAS** [comment]
@zoeyli-46 - I think what I'll do is leave this ticket as the "Review Changes" page and create a new ticket that is specific to the "Reason for Edits" that will essentially be a part 1 of this ticket.

**Comment by patrickisaac** [comment]
Refinement Notes:
- Dev checklist still required
- review code in CIF (might be helpful)

**Comment by patriciarussellCAS** [comment]
@ayeshmcg can you add which ticket is blocking this above so it's clear? Thanks

**Comment by ayeshmcg** [comment]
@patriciarussellCAS  this was blocked by https://github.com/bcgov/cas-reporting/issues/535

**Comment by zoeyli-46** [comment]
Hey @ayeshmcg looks great!
Some design notes to make it match the wireframes
- [ ] Add page title "Review changes" 
- [ ] Change title in the tasklist from change review to "Review changes"
- [ ] move the alert underneath review changes title
- [ ] make the font size of "reason for changes" match the section titles size

<img width="2174" height="1020" alt="Image" src="https://github.com/user-attachments/assets/57ca3d5e-8461-449e-8298-e9ad5379cc8d" />

Refer to the [wireframe here](https://www.figma.com/design/LsDmLDCdnJqI8UahyqvOD8/BCIERS-Designs?node-id=27418-5212&t=X788wHN76sNIiLa3-4)

<img width="614" height="371" alt="Image" src="https://github.com/user-attachments/assets/6bbf71b9-9696-40b0-b255-b754153b16eb" />

**Comment by fviduya** [comment]
I have so far tested these scenarios in DEV, here are the test documents for reference:
- [x] Verify that when registration purpose = [New Entrant Operation](https://buttoninc-my.sharepoint.com/:w:/g/personal/francesca_button_is/EcUgGaWPfYVMmeEf38jwT7QBc_AfrwRWVEMQqVsKdMBF1Q?e=MBnRQ4), fields for New Entrant display correctly - **NOTE: Fields for new entrant are showing in the Review Changes page, but they are not readable and seems to be in JSON format**
- [ ] Verify that when [registration purpose has changed](https://buttoninc-my.sharepoint.com/:w:/g/personal/francesca_button_is/ERXXOfy2AzRPpXk36652Y9cBPlHAn2oNjGLNO-7NqxmuZA?e=xLzlyZ), the difference between previous report version and current report version do not show on the Review Changes page - **NOTE: Currently shows**
- [ ] Given I am on the Review Changes page, I see the following:
   - [x] The page section and fields that were changed, in view-only mode
   - [x] The diff shows the old value with a strikethrough, and the new value in bold
   - [x] Changes to the report information fields will show all of the headers and subheaders that precede the field that was changed. i.e., if an emissions field was changed, then the facility name, the activity name, fuel, and emission title is displayed. Screenshot below as example.
   - [ ] `Review Changes` header
   - [ ] `Change review` is renamed to `Review Changes` in the task list
   - [ ] An alert `Your compliance obligation may be affected by changes to the values you have made.` is visible at the top of the page - **NOTE: Visible, but is at the bottom of the page**
   - [ ] The version number below the `Reason for change` section (based on the [wireframe](https://www.figma.com/design/LsDmLDCdnJqI8UahyqvOD8/BCIERS-Designs?node-id=27418-5212&t=eDv4xHOJv4qe5V8J-0))

Based on discussion earlier with @ayeshmcg and @zoeyli-46, the following changes are going to be implemented:
- Add `Review Changes` header at the top
- Move the alert message under the `Review Changes` header
- Rename `Change review` to `Review Changes`
- Add Production Data to Final Review page and Review Changes page
- Rename `Report Operation` to `Review Operation Information`
- In the Final Review page, change `Reason for Edits` to `Reason for Change`
- When facility is renamed, the old facility should have a strikethrough and new facility name should be in bold - currently when facility name is renamed, two sections of facility record is created in Review Changes (one shows as deleted with previous values, other one shows as added with new values)
- Fields that have values=0 should display as 0 instead of `null`
- The alert should be shown only when registration purpose = OBPS Regulated Operation or Opted-In Operation
- Change the size of the header of `Reason for Change` in the Review Changes page

Moving this back to In Progress for the additional changes. cc: @patriciarussellCAS

**Comment by ayeshmcg** [comment]
@fviduya  Added all the changes we discussed

**Comment by fviduya** [comment]
I have retested this in the DEV environment, here are the test documents:
- New Entrant Operation [test document](https://buttoninc-my.sharepoint.com/:w:/g/personal/francesca_button_is/EfNAw4QBQhNNq4DsG4-sfD0BlpsLvnVMwuT0_-mkLuFFLw?e=pkk4xe)
- Regulated Operation [test document](https://buttoninc-my.sharepoint.com/:w:/g/personal/francesca_button_is/Eb1WJIZq9NpKpeSnU_kwOooB5Qr5u9h3Qsa39NghQwbFgA?e=v72oCt)
- Opted-in Operation [test document](https://buttoninc-my.sharepoint.com/:w:/g/personal/francesca_button_is/ESZKKeW0P8ZNv0ILsSfOZr0BA_5YySqwlFoXN4YWXB9JRw?e=AmeErB)

**NOTE: these documents are huge, I've put up a table of contents there. But if you're in the middle of the document it might get messy, you can open navigation on the side (Click View at the top > Navigation)**

From the ACs above, I can verify that:
- [x] Given I am on the Review Changes page, I see the following:
   - [ ] the page section and fields that were changed, in view-only mode - **NOTE: For New Entrant Operation and Opted-in Operation, it looks like the changes for the activity page is missing. I've added the link to the document above, there's also a table of contents in the document. Please select Activity pages under Review changes in the table of contents to get to the screenshot.**
   - [x] The diff shows the old value with a strike through, and the new value in bold
   - [x] Changes to the report information fields will show all of the headers and subheaders that precede the field that was changed. i.e., if an emissions field was changed, then the facility name, the activity name, fuel, and emission title is displayed. Screenshot below as example.
- [x] `Review Changes` header is visible at the top when user is in the Review Changes page
- [x] The alert message is visible at the bottom of the `Review Changes` header
- [x] `Change Review` has been renamed to `Review Changes` in the task list
- [x] `Report Operation` has been renamed to `Review Operation Information` in the Review Changes page
- [x] `Reason for Edits` has been renamed to `Reason for Change` in the Final Review page
- [x] When facility is renamed, the old facility name has a strikethrough and new facility name is written in bold face
- [ ] Fields that have a value of 0 is displayed as 0 instead of `null` - **NOTE: Still displays as `null` in the Additional Reporting Data, the other fields show 0**
- [x] The alert `Your compliance obligation may be affected by changes to the values you have made.` is shown only when registration purpose = OBPS Regulated Operation or Opted-in Operation
- [x] The size of the header of `Reason for Change` has been resized to follow the other headers in the Review Changes page

Hi @patriciarussellCAS and @ayeshmcg, these are the following I noticed that might be a higher priority:
- During testing for New Entrant Operation and Opted-in Operation, the changes I made for the activity pages (and everything under Report Information) was not visible in the Review Changes page, it shows `deleted`. I'm not certain if it is because I did not change the facility name, cc: @ayeshmcg 
- During testing for Regulated Operation, I completely changed the product, but the change is not visible in the Review Changes page. It shows no changes for that header.

**Comment by ayeshmcg** [comment]
@fviduya There were some breaking changes when we updated deepDiff, We have a PR up for fixing those(https://github.com/bcgov/cas-registration/pull/3757).

**Comment by ayeshmcg** [comment]
@fviduya https://github.com/bcgov/cas-registration/issues/3747 has been merged

**Comment by fviduya** [comment]
Thanks, @ayeshmcg! I will retest once the code is deployed 😄

**Comment by fviduya** [comment]
I have retested this in the DEV environment, here are the test documents for reference:
- [New Entrant Operation](https://buttoninc-my.sharepoint.com/:w:/g/personal/francesca_button_is/EZws1rE997xKl_TlOSYgF1EBpSUN1eevTjS9GE7B-AgsMw?e=lxyKDE)
- [Changing Registration Purpose](https://buttoninc-my.sharepoint.com/:w:/g/personal/francesca_button_is/EbCixZ2kIWxLjSK-vVuMlZUBUgyFO5XbEZZxdFaO-2AQ4g?e=UVevQT) (regulated to reporting operation)
- [Electricity Import Operation](https://buttoninc-my.sharepoint.com/:w:/g/personal/francesca_button_is/EaRrQKy6ocVFnvhGyYUGWK0B6Qaij5_GaZHXSW0dz2OQdA?e=c7sv6q)
- [Regulated Operation](https://buttoninc-my.sharepoint.com/:w:/g/personal/francesca_button_is/EXlXNXzRrkNJtDM-RLmO_X0BezEKE8fikW4M-p7vIW8WEg?e=oDudJV)

As discussed with @patriciarussellCAS and @ayeshmcg earlier, these were the issues found during testing:
- Under New Entrant Operation document:
   - When Production Data is deselected, the review changes page shows `No changes found` - **No issue if user changes the value instead, this could be an edge-case scenario**
   - Bottom 3 values of New Entrant Information in the Final Review page has incorrect values - #876 
   - There is a missing category (or a mislabel) in the Allocation of Emissions section
   - The value for `did your non-attributable emissions exceed 100 tCO2e?` in Non-attributable emissions did not change to `No` - **Shows `No` in the Final Review page**

- Under Electricity Import Operation document:
   - When user deletes an operation representative in the Review Operation Information page, the deleted name is still visible both in Review Changes page and Final Review page - #877 

I can verify that: 
- [x] Fields that have a value of 0 is displayed as 0 instead of null
- [x] In the Review changes page, I can see the page section and fields that were changed, in view-only mode

Moving to Pending PO for review of test and notes. Hi @patriciarussellCAS, please let me know if you need assistance in reviewing the test documents. Thank you!

**Comment by patriciarussellCAS** [comment]
Thanks @fviduya - I have tested a Regulated LFO, a Reporting SFO and a New Entrant LFO. 
I can confirm that the Production Data changes not displaying on the Review Changes page is also occurring for the regulated LFO, so not just New Entrants. The added or removed production information does show up on the final review page though. 
I'll make a separate ticket for this and accept this one. Thanks.

**Comment by patriciarussellCAS** [comment]
I have reviewed this feature in TEST and have the following notes: 
1. I found a number of bugs but I believe they are separate from this feature so will make separate notes/tickets for those
2. Overall this works as expected for Regulated, Reporting, New Entrant, EIO and Opt-ins (though noting the bugs already addressed in Chesca's post above)
3. Testing a Reporting Operation - I added an activity on the Review Operation Info Page. This shows up on the Review changes page, but doesn't have the "added" label, or show any of the fields. I think this is ok because we see the data fields on the Final Review page, but it seems a bit inconsistent with how we show need source types, units, emissions etc. that get added. 

<img width="1086" height="715" alt="Image" src="https://github.com/user-attachments/assets/3364518a-6fe0-46bf-af4b-379f9fa5e353" />

<img width="1177" height="742" alt="Image" src="https://github.com/user-attachments/assets/161790a1-83ae-4709-967d-53b4b2a8faba" />

4. When I remove an activity from the 'Review Operation Info' page, for an LFO, this does not update in the facility reports, so that activity is still listed in the task list and checked off in the Review Facility page. On the Review changes page, it shows as removed from the Review Operation Info page, but the activity is still there for the emissions reports. I assume this means that LFO users will need to both remove the activity from the Review Op Info page AND uncheck the activity on the Review Facilities form to fully remove it.  - Note: I also added a new facility to this LFO. The removed activity was NOT selected or visible for this facility on the review facility info page, so the update I made to the Review Op Info page DID carry forward to this new facility, but not to the 2 facilities already in there. This probably needs a separate ticket and investigation. I don't think it should stop the release. 

5. Not sure how we deal with this situation: In a supplementary report, I make a bunch of changes to emissions data, and then go all the way to Final Review page, but before I submit, I go back and make NEW changes to emissions data - even though my updates are saved and reflected properly in the final review page, these new changes don't show up in the Review Changes page. I assume this is because it's only showing the diff between a submitting report and the next submitted report. Again, something I think we need to investigate separately. 
Here's some screenshots from the scenario - New Entrant, changes my emissions, which impacted my allocation numbers, but these changes don't show up in the review changes page. The only changes that display are the methodology and total emissions

First allocation page in my supp report: 
<img width="463" height="317" alt="Image" src="https://github.com/user-attachments/assets/04fa4e5d-5d20-4287-bfaf-b2dc1b170b14" />

Second allocation after going back and changing activity data:
<img width="447" height="302" alt="Image" src="https://github.com/user-attachments/assets/4d92f239-1db3-4b17-97b2-c78c634c15b9" />

Review Changes Page: 
<img width="349" height="221" alt="Image" src="https://github.com/user-attachments/assets/12cbd4c4-ac96-4773-b2f4-d58ea227fb9d" />

**Comment by TammieV-BCG** [comment]
Tested in UAT by BA. 
Review changes page returns an error message for Reporting Operation, Regulated Operation and EIO didnt test other registration purposes.
 see defect report [here](https://bcgov.sharepoint.com/:w:/r/teams/00605/Shared%20Documents/General/BCIERS%20Testing/UAT%20Defect%20Reports/Post%20Launch%20UAT/Defect%20Report-BO_%20479%26877.docx?d=w98e8e2ecbe91424fb043684a38354247&csf=1&web=1&e=zEUfUz)

<img width="1616" height="754" alt="Image" src="https://github.com/user-attachments/assets/50d9974a-4665-48f5-ab0f-44c73d4a6b4e" />

**Comment by patriciarussellCAS** [comment]
Adding my notes from testing: 
I can confirm that this is working properly for: 

- [x] New Entrant LFO
- [x] Regulated Operation SFO
- [x] EIO
- [x] Opted-in 
- [x] Changing registration purpose
- [x] Regulated Operation LFO
- [x] Reporting Operation LFO
- [x] Reporting Operation SFO
